### PR TITLE
Adds support for failures_before_warning and adds failues_before options on Nomad client service

### PR DESCRIPTION
**Pull request by mikenomitch**
This PR does 2 things:

First, checks in Nomad now can have a failures_before_warning value similar to failures_before_critical except the check changes to a warning before going red.

To test manually, deploy [this job](https://gist.github.com/mikenomitch/654dc9f5bbdefc747eddaaa7ba1ff207), then kill Nomad, then kill the docker container manually, You will see the check turn into a warning and then eventually go red.

Secondly, failures_before_warning and failures_before_critical can now be set on the Nomad services in the Nomad agent config. This is actually what prompted the first change.

To test manually, set this as your Nomad agent config, then kill Nomad and watch the Nomad checks go to yellow then red, but not immediately.

```
consul {
    client_failures_before_critical = 4
    client_failures_before_warning = 2
    server_failures_before_critical = 2
    server_failures_before_warning = 1
}
```

### CODE REVIEW

**Comment by mikenomitch** [comment]
@atirabassi-eb, I think this is roughly the change you would need for the request you had. While I was in the space, I just threw up a draft PR to then go talk to the team about. If its an easy enough fix, we can hopefully get it in ourselves. If it involves a lot more, then I might see if you can take the lead on those changes.

**Comment by mikenomitch** [comment]
Closes https://github.com/hashicorp/nomad/issues/19028

File: command/agent/command.go
Comment (tgross) [inline command/agent/command.go:1577]: Similar to the `-consul-client-http-check-name` field above, we should make it clear in the docstring that this is the Consul health check for the Nomad client only. (Because Consul itself has a client and that just makes the flag slightly ambiguous.)

File: website/content/docs/commands/agent.mdx
Comment (tgross) [inline website/content/docs/commands/agent.mdx:72]: Having the little empty `[]` lets a lot of commonly used tools pick this up as needing syntax highlighting:

```suggestion
- `-consul-client-failures-before-critical=<num>`: Equivalent to the
  [client_failures_before_critical][] config option.

- `-consul-client-failures-before-warning=<num>`: Equivalent to the
  [client_failures_before_warning][] config option.
```

File: command/agent/agent.go
Comment (tgross) [inline command/agent/agent.go:1190]: Why just clients?

File: website/content/docs/job-specification/check.mdx
Comment (tgross) [inline website/content/docs/job-specification/check.mdx:473]:
```suggestion
[consul_failure_before_critical]: /consul/api-docs/agent/check#failuresbeforecritical
[consul_failure_before_warning]: /consul/api-docs/agent/check#failuresbeforewarning
```

File: nomad/structs/services.go
Comment (tgross) [inline nomad/structs/services.go:517]: Should this be `failures` or some unique name?

**Review by tgross — COMMENTED** [review]
Looks great @mikenomitch!

I've left some comments / questions to fix up. You'll also need to run `make cl` but this PR should get two changelog entries because it's two different features. And then once you've got it merged don't forget to add the backport label to 1.7.x so it can go out in the next 1.7.x release!

File: nomad/structs/services.go
Comment (mikenomitch) [reply nomad/structs/services.go:517]: Changing - I am leaving the existing string as "failures" in case this hash is cached somewhere before the user updates and then on update a new hash appears - Does that matter? I'm not sure exactly where the hash is used. If its just thrown away then I can update both strings.

File: website/content/docs/commands/agent.mdx
Comment (mikenomitch) [reply website/content/docs/commands/agent.mdx:72]: Is this correct? I see all the other links without this?

**Comment by mikenomitch** [comment]
@tgross thanks for the review. I think this is ready for another look. - I added the server checks in here as well. - A couple questions for you in the inline comments.

File: website/content/docs/commands/agent.mdx
Comment (tgross) [reply website/content/docs/commands/agent.mdx:72]: Oh sorry maybe that was confusing. It doesn't impact the links on the actual website at all!

But depending on what editor / LSP / etc you use, having this trailing `[]` can make the difference between the `[client_failures_before_warning]` being shown as plain text vs highlighted as a link when you're editing the markdown. (Emacs + `markdown-mode` is one example, I know personally, but @angrycub has pointed out other cases I can't recall.)

File: nomad/structs/services.go
Comment (tgross) [reply nomad/structs/services.go]: Yeah, the hash gets used as part of the ID we register with Consul. We then use that to clean up the checks on job stop. If we change the hash without the user having submitted a jobspec change to the service/check, we end up leaking the check if you upgrade the Nomad client in place -- the service client will have the wrong check ID.

So changing to `"failures-before-warning"` but not touching the other string is good (as you've done in your edit). That'll preserve the existing hash.

File: command/agent/command.go
Comment (tgross) [inline command/agent/command.go:1611]:
```suggestion
    Consul health check shows a warning. Defaults to 0.
```

File: command/agent/command.go
Comment (tgross) [inline command/agent/command.go:1581]:
```suggestion
    Consul health check shows a warning. Defaults to 0.
```

File: .changelog/19336.txt
Comment (tgross) [inline .changelog/19336.txt:3]: This needs a second entry for the jobspec change (just a second `release-note:improvement` block under this one).

File: website/content/docs/configuration/consul.mdx
Comment (tgross) [inline website/content/docs/configuration/consul.mdx:199]:
```suggestion
  consecutive failures before the Nomad server Consul health check shows a warning.
```

**Review by tgross — APPROVED** [review]
LGTM once the changelog entry and a couple typos are fixed. Don't forget to squash-merge!

**Comment by mikenomitch** [comment]
Bah, sorry about the slop! - On it now

File: .changelog/19336.txt
Comment (mikenomitch) [reply .changelog/19336.txt:3]: Ah dang didn't realize how this worked -  I reran the "make cl" command twice thinking I could add 2 changelog entries, but I guess it overwrote the initial.

**Pull request merged**
